File: retrofit/src/main/java/com/kcst/retrofit/PostRequestHandler.java
package com.kcst.retrofit;

import androidx.constraintlayout.widget.ConstraintSet;

import com.google.gson.Gson;

import java.io.IOException;
import java.io.StringReader;
import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Objects;
import java.util.Properties;

import retrofit2.Call;
import retrofit2.Response;

public class PostRequestHandler implements RequestHandler {

    @Override
    public <T extends BaseResponse> T requestHandler(RetrofitService retrofitService,
                                        BaseRequest baseRequest,
                                        Class<T> responseClazz,
                                        Gson gson) throws IllegalAccessException, InstantiationException, NoSuchMethodException, InvocationTargetException {

        if (baseRequest.headers == null || baseRequest.headers.isEmpty()) {
            baseRequest.headers = new HashMap<>();
            baseRequest.headers.put("Content-Type", "application/x-www-form-urlencoded");
        }
        HashMap<String, String> requestParams = new HashMap<>();
        if (baseRequest.getData() != null) {
            Class<?> aClass = baseRequest.getData().getClass();
            Field[] declaredFields = aClass.getDeclaredFields();
            Constructor declaredConstructor = aClass.getDeclaredConstructor();
            declaredConstructor.setAccessible(true);
            for (Field field : declaredFields) {
                field.setAccessible(true);
                requestParams.put(field.getName(), Objects.requireNonNull(field.get(baseRequest.getData())).toString());
            }
        }
        T baseResponse;
        try {
            Call<String> post = retrofitService.post(baseRequest.headers, baseRequest.path, requestParams);
            Response<String> response = post.execute();
            if (response.isSuccessful()) {
                Type genericSuperclass = responseClazz.getGenericSuperclass();
                baseResponse = gson.fromJson(new StringReader(response.body()),genericSuperclass);
            } else {
                baseResponse = responseClazz.newInstance();
                baseResponse.isSuccess=false;
                baseResponse.errorCode=response.code();
                baseResponse.errorMsg=response.message();
            }
            return baseResponse;
        } catch (Exception e) {
            baseResponse = responseClazz.newInstance();
            baseResponse.isSuccess=false;
            baseResponse.errorCode=501;
            baseResponse.errorMsg=e.getMessage();
            return baseResponse;
        }

    }
}

File: retrofit/src/main/java/com/kcst/retrofit/PostRequestHandler.kt
package com.kcst.retrofit

import kotlin.reflect.full.declaredMemberProperties
import kotlin.reflect.jvm.isAccessible

class PostRequestHandler : RequestHandler {
    override suspend fun <T : BaseResponse<*>> requestHandler(
        retrofitService: RetrofitService,
        baseRequest: BaseRequest,
        responseClazz: Class<T>,
    ): T {
        if (baseRequest.headers == null) {
            baseRequest.headers = mutableMapOf<String, String>().apply {
                put("Content-Type", "application/x-www-form-urlencoded")
            }
        }
        val requestParams = mutableMapOf<String, String>()
        val requestKClass = baseRequest::class
        requestKClass.declaredMemberProperties.forEach {
            it.isAccessible = true
            requestParams[it.name] = it.call(baseRequest).toString()
        }
        var response: T = responseClazz.newInstance()
        try {
            val call =
                retrofitService.post(baseRequest.headers, baseRequest.getPath(), requestParams)
            val strResponse = call.execute()
            if (strResponse.isSuccessful) {
                response =
                    GsonUtil.fromJson(strResponse.body()!!, responseClazz)
            } else {
                response.isSuccess = false
                response.errorCode = strResponse.code()
                response.errorMsg = strResponse.message()
            }
        } catch (e: Exception) {
            response.isSuccess = false
            response.errorCode = 501
            response.errorMsg = e.message
        }
        return response
    }

}